examples/timebase_master.py

```diff
@@ -0,0 +1,148 @@
+                pos.beat += 1
+
+                if pos.beat > pos.beats_per_bar:
+                    pos.beat = 1
+                    pos.bar += 1
+                    pos.bar_start_tick += pos.beats_per_bar * pos.ticks_per_beat
+
+                    if self.debug:
+                        print("Pos:", jack.position2dict(pos))
+
+    def become_timebase_master(self, conditional=None):
+        return self.set_timebase_callback(self._tb_callback, conditional
+                                          if conditional is not None
+                                          else self.conditional)
+
+
+def main(args=None):
+    ap = argparse.ArgumentParser(description=__doc__)
+    ap.add_argument(
+        '-d', '--debug',
+        action='store_true',
+        help="Enable debug messages")
+    ap.add_argument(
+        '-c', '--conditional',
+        action='store_true',
+        help="Exit if another timebase master is already active")
+    ap.add_argument(
+        '-n', '--client-name',
+        metavar='NAME',
+        default='timebase',
+        help="JACK client name (default: %(default)s)")
+    ap.add_argument(
+        '-m', '--meter',
+        default='4/4',
+        help="Meter as <beats-per-bar>/<beat-type> (default: %(default)s)")
+    ap.add_argument(
+        '-t', '--ticks-per-beat',
+        type=int,
+        metavar='NUM',
+        default=1920,
+        help="Ticks per beat (default: %(default)s)")
+    ap.add_argument(
+        'tempo',
+        nargs='?',
+        type=float,
+        default=120.0,
+        help="Tempo in beats per minute (default: %(default)s)")
+
+    args = ap.parse_args(args if args is not None else sys.argv[1:])
+
+    try:
+        beats_per_bar, beat_type = (int(x) for x in args.meter.split('/', 1))
+    except (TypeError, ValueError):
+        print("Error: invalid meter: %s\n" % args.meter)
+        ap.print_help()
+        return 2
+
+    try:
+        tbmaster = TimebaseMasterClient(
+            args.client_name,
+            bpm=max(0.0, min(300.0, args.tempo)),
+            beats_per_bar=beats_per_bar,
+            beat_type=beat_type,
+            ticks_per_beat=args.ticks_per_beat,
+            debug=args.debug)
+    except jack.JackError as exc:
+        return "Could not create timebase master JACK client: %s" % exc
+
+    with tbmaster:
+        if tbmaster.become_timebase_master(args.conditional):
+            try:
+                print("Press Ctrl-C to quit...")
+                tbmaster.stop_event.wait()
+            except KeyboardInterrupt:
+                print('')
+            finally:
+                try:
+                    tbmaster.release_timebase()
+                except jack.JackError:
+                    # another JACK client might have grabbed timebase master
+                    pass
+        else:
+            return "Timebase master already present. Exiting..."
+
+
+if __name__ == '__main__':
+    sys.exit(main(sys.argv[1:]) or 0)
```

Comment:
Since this is the default behavior, you don't really have to pass anything to `main()`, right?

Reply:
I use the following strategy for all my command line scripts (imports not shown):

```py
def main(args=None):
    ap = argparse.ArgumentParser()
    args = ap.parse_args(args if args is not None else sys.argv[1:])

if __name == '__main__':
    sys.exit(main(sys.argv[1:]) or 0)
```

This allows the following:

* You can call the script directly with `python myscript.py` (args are passed via the call to `main()`).
* You can use the script as a `console_script` entry point (`console_scripts = ['myscript = mypkg.myscript:main]`) (`main()` is called without arguments and the command line args have to be retrieved via `sys.argv`).
* You can `import` the script in another Python file and call the `main()` function manually, passing the command line as a string. 
* The `sys.exit(main(...) or 0)` allows the `main()` function to just return an exit code or an error string to set the program exit code or print an error code to `stderr`, but when you call `main()` manually, you can still use the return value (`ArgumentParser.parse_args()` may still terminate the program, so this isn't perfect). Returning `None` (the default with no `return` statement), sets the exit code to 0 (success).

Reply:
As discussed above `ap.parse_args(args if args is not None else sys.argv[1:])` can be shortened to `ap.parse_args(args)`.

Reply:
I think there are two different things:

* calling `main()` vs `main(sys.argv[1:])` (your first 3 points)
* terminating on error or returning something (your last point)

AFAICT, there is no disadvantage in calling `main()` without arguments here, it works exactly the same in all mentioned cases.

Your last point is a separate thing.

I guess it's possible to do it that way, but I'm not sure if you are gaining anything. It probably even makes things worse.

On one hand, it might be a YAGNI situation just adding useless noise.
Did you ever hear of a usage where this mattered?

On the other hand, I'm not sure what you would even do when you get an "error value" except for terminating the program.
Is there a conceivable situation where you wouldn't want to terminate the program?

And even if there is a very teeny tiny chance that somebody reeeeally doesn't want to terminate, they could still catch `SystemExit`, right?

> (`ArgumentParser.parse_args()` may still terminate the program, so this isn't perfect)

Exactly, far from it.

It's actually an extremely common case.

And I think this is the main argument here.
Why should a wrong `--tempo` argument take a completely different code path than a wrong `--meter` argument?

All in all, it sounds like much noise for little gain, doesn't it?